cognite/client/utils/_pyodide_helpers.py

```diff
@@ -0,0 +1,218 @@
+        action="ignore",
+        category=UserWarning,
+        message="Your installation of 'protobuf' is missing compiled C binaries",
+    )
+    # - AssetsAPI.create_hierarchy is not implemented yet (custom threading solution)
+    cc._api.assets._AssetPoster = NotImplementedAssetPoster  # type: ignore [assignment, misc]
+
+    # - DatapointsAPI.retrieve/retrieve_arrays/retrieve_dataframe:
+    cc._api.datapoints.PriorityThreadPoolExecutor = SerialPriorityThreadPoolExecutor  # type: ignore [assignment, misc]
+    cc._api.datapoints.as_completed = serial_as_completed  # type: ignore [assignment]
+
+
+def http_client__init__(
+    self: HTTPClient,
+    config: HTTPClientConfig,
+    session: Session,
+    retry_tracker_factory: Callable[[HTTPClientConfig], _RetryTracker] = _RetryTracker,
+) -> None:
+    import pyodide_http  # type: ignore [import]
+
+    self._old__init__(config, session, retry_tracker_factory)  # type: ignore [attr-defined]
+    self.session.mount("https://", pyodide_http._requests.PyodideHTTPAdapter())
+    self.session.mount("http://", pyodide_http._requests.PyodideHTTPAdapter())
+
+
+class NotImplementedAssetPoster:
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        raise NotImplementedError("AssetsAPI.create_hierarchy is not pyodide/web-browser compatible yet! Stay tuned!")
+
+
+class DummyFuture(TaskFuture):
+    def __init__(self, fn: Callable[..., T_Result], *args: Any, **kwargs: Any):
+        self._task = functools.partial(fn, *args, **kwargs)
+        self._result: Optional[T_Result] = None
+        self._is_cancelled = False
+
+    def result(self) -> T_Result:
+        if self._is_cancelled:
+            raise CancelledError
+        if self._result is None:
+            self._result = self._task()
+        return self._result
+
+    def cancel(self) -> None:
+        self._is_cancelled = True
+
+
+def serial_as_completed(dct: Dict[DummyFuture, Any]) -> Iterator[DummyFuture]:
+    return iter(dct.copy())  # Will raise StopIteration when done (we want this):
+
+
+class SerialPriorityThreadPoolExecutor(TaskExecutor):
+    def __init__(self, max_workers: int = None) -> None:
+        if max_workers != 1:
+            raise RuntimeError("Max_workers must be == -e^(i*pi), or 1 if you wondered")
+
+    def submit(self, fn: Callable[..., T_Result], *args: Any, **kwargs: Any) -> DummyFuture:
+        if "priority" in inspect.signature(fn).parameters:
+            raise TypeError(f"Given function {fn} cannot accept reserved parameter name `priority`")
+        kwargs.pop("priority", None)
+        return DummyFuture(fn, *args, **kwargs)
+
+    def shutdown(self, wait: bool = False) -> None:
+        return None
+
+    def __enter__(self) -> SerialPriorityThreadPoolExecutor:
+        return self
+
+    def __exit__(
+        self,
+        type: Optional[type[BaseException]],
+        value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        self.shutdown()
```

Comment:
I would suggest moving these into the concurrency module along with the PriorityThreadPoolExecutor, and toggling their usage with the same ConcurrencySettings option. Makes it easier to follow if it's done the same way for the different implementations.

Reply:
Moved what could be moved without too much trouble. Patching now looks unified:

```py
# - Set all usage of thread pool executors to use dummy/serial-implementations:
cc.utils._concurrency.ConcurrencySettings.executor_type = "mainthread"
cc.utils._concurrency.ConcurrencySettings.priority_executor_type = "mainthread"
```

Skipped: Unable to use the same singleton pattern due to various logic calling on the pool to shut down explicitly. It seems like a significant risk to change that logic atm.

Reply:
Also moved the usage of `concurrent.future.as_completed` to a pool method, that way we can forget about patching it 😄